NLSY/nlsy_lib.R

```diff
@@ -248,6 +251,191 @@ nlsy_get_highest_grade_completed_df = function()
+            starts_with('CV_BIO_CHILD_HH'),
+            -contains('U18') #Dropping variables with only children under 18
+
+        ) |>
+        pivot_longer(
+            starts_with('CV_BIO_CHILD_HH'),
+            names_to='year',
+            names_pattern='CV_BIO_CHILD_HH_(\\d\\d\\d\\d)',
+            values_to='bio_child_hh',
+            values_transform = list(bio_child_hh = as.integer)
+        )
+
+    data2 = nlsydf |>
+        select(
+            id =        "PUBID_1997",
+            starts_with('CV_BIO_CHILD_NR'),
+            -contains('U18') #Dropping variables with only children under 18
+
+        ) |>
+        pivot_longer(
+            starts_with('CV_BIO_CHILD_NR'),
+            names_to='year',
+            names_pattern='CV_BIO_CHILD_NR_(\\d\\d\\d\\d)',
+            values_to='bio_child_nr',
+            values_transform = list(bio_child_nr = as.integer)
+        )
+
+    data = left_join(data1, data2, by=c('id', 'year'))|>
+        mutate(year = as.integer(year))|>
+        group_by(id) |>
+        complete(year=1997:2019) |>
+        fill(bio_child_hh, bio_child_nr)|>
+        mutate(bio_child_hh = case_when(
+                   is.na(bio_child_hh) ~ 0, # Assuming zeroes are part of the survey auto skips
+                   TRUE ~ bio_child_hh),
+               bio_child_nr = case_when(
+                   is.na(bio_child_nr) ~ 0, # Assuming zeroes are part of the survey auto skips
+                   TRUE ~ bio_child_nr)
+        )|>
+        mutate(bio_child = bio_child_hh + bio_child_nr)|>
+        select(-bio_child_nr, -bio_child_hh, -bio_child)
+
+return(data)
+}
+
+
+#' Adds longitudinal income and wealth vars to the input dataset
+nlsy_get_income_wealth_df = function(){
+    income <- nlsydf |>
+        select(id = PUBID_1997, income_source = CV_HH_INCOME_SOURCE_1997, matches('CV_INCOME_GROSS_YR|CV_INCOME_FAMILY')) |>
+        pivot_longer(-c('id', 'income_source'),
+                     names_to = 'year', values_to = 'income') |>
+        mutate(year = as.numeric(str_sub(year, start=-4)))
+    networth_97_03 <- nlsydf |>
+        select(id = PUBID_1997, matches('CV_HH_NET_WORTH_Y')) |>
+        pivot_longer(-id,
+                     names_to = 'year', values_to = 'networth') |>
+        mutate(year = as.numeric(str_sub(year, start=-4)))
+
+    networth_5yr <- nlsydf |>
+        select(id = PUBID_1997, age_resp = "CV_AGE_12/31/96_1997", matches('CVC_HH_NET_WORTH_\\d{2}')) |>
+        pivot_longer(-c('id', 'age_resp'),
+                     names_to = 'year', values_to = 'networth') |>
+        mutate(year = as.numeric(str_sub(year, start=-7, end=-6)), # Get age for 5-yr wealth vars
+               year = 1997 + (year - age_resp)) # Add correct number of years based on resp's starting age
+
+    networth_merged <- tidylog::full_join(networth_97_03, networth_5yr, by = c('id', 'year', 'networth')) |>
+        select(-age_resp)
+
+    dups <- count(networth_merged, id, year) |> filter(n>1)
```

Comment:
I have not been checking for duplicates. Did you find any here? If so I am curious what caused them and whether we should do this check elsewhere.

Reply:
Basically, I'm using the respondent's original age to infer which year the 5-year wealth questions were asked to them. If that year happens to fall within 1997-2003, then there are technically duplicate ID-years. But looking at the data, whenever this happens the 1997-2003 value is always missing so I don't think it's an issue.